metagenomics.py

```diff
@@ -1145,6 +1150,122 @@ def kraken_library_ids(library):
+    )
+    util.cmd.common_args(parser, (('loglevel', None), ('version', None), ('tmp_dir', None)))
+    util.cmd.attach_main(parser, filter_bam_to_taxa, split_args=True)
+    return parser    
+
+def filter_bam_to_taxa( in_bam, 
+                        in_reads_to_tax_ID_map_file,
+                        out_bam,
+                        in_taxdb_nodes_path, 
+                        in_taxdb_names_path, 
+                        tax_names=None, 
+                        tax_ids=None,
+                        omit_children=False,
+                        read_id_col=1, 
+                        tax_id_col=2,
+                        JVMmemory=None ):
+    """
+        Filter an (already classified) input bam file to only include reads that have been mapped to specified 
+        taxonomic IDs or scientific names. This requires a classification file, as produced 
+        by tools such as Kraken, as well as the NCBI taxonomy database.
+    """
+    tax_ids = set(tax_ids) if tax_ids else set()
+    tax_names = tax_names or []
+    # use TaxonomyDb() class above and tree traversal/collection functions above 
+    db = TaxonomyDb(nodes_path=in_taxdb_nodes_path, names_path=in_taxdb_names_path, load_nodes=True, load_names=True)
+    db.children = parents_to_children(db.parents)
+
+    paired_read_base_pattern = re.compile(r'^(.*)/[1-2]$')
+
+    # get taxIDs for each of the heading values specifed (exact matches only)
+    tax_ids_from_headings = set()
+    for heading in tax_names:
+        # map heading to taxID
+        name_pattern = re.compile(heading.lower())
+        for row_tax_id, names in db.names.items():
+            found_heading = False
+            if type(names) == list:
+                # if taxID->list (of names)
+                for name in names:
+                    if name_pattern.match(name.lower()):
+                        tax_ids_from_headings.add(row_tax_id)
+                        log.debug("Found taxName match: %s -> %s" % (row_tax_id,name))
+                        found_heading = True
+                        break
+                if found_heading:
+                    break
+            else:
+                # if taxID->str
+                if name_pattern.match(names.lower()):
+                    tax_ids_from_headings.add(row_tax_id)
+                    log.debug("Found taxName match: %s -> %s" % (row_tax_id,names))
+                    break
+
+    tax_ids |= tax_ids_from_headings
+
+    log.debug("tax_ids %s" % tax_ids)
+    log.debug("tax_names %s" % tax_names)
+
+    # extend tax_ids to include IDs of children
+    tax_ids_to_include = set()
+    for tax_id in tax_ids:
+        tax_ids_to_include.add(tax_id)
+        if not omit_children:
+            child_ids = collect_children(db.children, set([tax_id]))
+            tax_ids_to_include |= set(child_ids)
+
+    tax_ids_to_include = frozenset(tax_ids_to_include) # frozenset membership check slightly faster
+
```

Comment:
I removed such a line since the set can be huge depending on the number of children (ex. `--taxNames="Viruses"`), and since we usually set `--loglevel=DEBUG` on DNAnexus it's nice to avoid filling the logfile with tax IDs.